File: 0110-balanced-binary-tree/0110-balanced-binary-tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    bool isBalanced(TreeNode* root) {
        if (!root) {
            return true;
        }

        int leftHeight = subtreeHeight(root->left);
        int rightHeight = subtreeHeight(root->right);
        bool compare = abs(leftHeight - rightHeight) <= 1;

        return isBalanced(root->left) && isBalanced(root->right) && compare;
    }
private:
    int subtreeHeight(TreeNode* node) {
        if (!node) {
            return 0;
        }

        return 1 + max(subtreeHeight(node->left), subtreeHeight(node->right));
    }
};

File: 0110-balanced-binary-tree/0110-balanced-binary-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def isBalanced(self, root: TreeNode | None) -> bool:
        height = self.depth(root)
        if height == -1:
            return False
        
        return True
 
    def depth(self, node) -> int:
        if not node:
            return 0
        
        leftDepth = self.depth(node.left)
        if (leftDepth == -1):
            return -1
        rightDepth = self.depth(node.right)
        if (rightDepth == -1):
            return -1

        if (abs(rightDepth - leftDepth) > 1):
            return -1
        
        return 1 + max(leftDepth, rightDepth)
        